backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisClusterExtension.java

```diff
@@ -0,0 +1,180 @@
+import com.google.inject.AbstractModule;
+import com.google.inject.Module;
+import com.google.inject.Provides;
+
+import scala.jdk.javaapi.OptionConverters;
+
+public class RedisClusterExtension implements GuiceModuleTestExtension {
+
+    public static class RedisClusterContainer extends ArrayList<GenericContainer> {
+        public RedisClusterContainer(Collection<? extends GenericContainer> c) {
+            super(c);
+        }
+
+        public RedisConfiguration getRedisConfiguration() {
+            return RedisConfiguration.from(this.stream()
+                    .map(redisURIFunction())
+                    .map(URI::toString)
+                    .toArray(String[]::new),
+                true, OptionConverters.toScala(Optional.empty()), OptionConverters.toScala(Optional.empty()));
+        }
+
+        public void pauseOne() {
+            GenericContainer firstNode = this.get(0);
+            firstNode.getDockerClient().pauseContainerCmd(firstNode.getContainerId()).exec();
+        }
+
+        public void unPauseOne() {
+            GenericContainer firstNode = this.get(0);
+            if (TRUE.equals(firstNode.getDockerClient().inspectContainerCmd(firstNode.getContainerId())
+                .exec()
+                .getState()
+                .getPaused())) {
+                firstNode.getDockerClient().unpauseContainerCmd(firstNode.getContainerId()).exec();
+            }
+        }
+    }
+
+    public static final Function<String, GenericContainer> redisContainerSupplier = alias ->
+        new GenericContainer<>(DEFAULT_IMAGE_NAME)
+            .withExposedPorts(DEFAULT_PORT)
+            .withCreateContainerCmdModifier(createContainerCmd -> createContainerCmd.withName("james-" + alias + "-test-" + UUID.randomUUID()))
+            .withCommand("redis-server", "/usr/local/etc/redis/redis.conf")
+            .withNetworkAliases(alias)
+            .withClasspathResourceMapping("redis_cluster.conf",
+                "/usr/local/etc/redis/redis.conf",
+                BindMode.READ_WRITE)
+            .waitingFor(Wait.forLogMessage(".*Ready to accept connections.*", 1)
+                .withStartupTimeout(Duration.ofMinutes(2)));
+
+    static final GenericContainer redis1 = redisContainerSupplier.apply("redis1");
+    static final GenericContainer redis2 = redisContainerSupplier.apply("redis2");
+    static final GenericContainer redis3 = redisContainerSupplier.apply("redis3").dependsOn(redis1, redis2);
+
+    private RedisClusterContainer redisClusterContainer;
+    private final Network network;
+
+    public RedisClusterExtension() {
+        this(Network.newNetwork());
+    }
+
+    public RedisClusterExtension(Network network) {
+        this.network = network;
+        redis1.withNetwork(network);
+        redis2.withNetwork(network);
+        redis3.withNetwork(network);
+    }
+
+    @Override
+    public void beforeAll(ExtensionContext extensionContext) throws IOException, InterruptedException {
+        redis3.start();
```

Comment:
Depend on would help start redis 1 and 2 also?

Reply:
yes,